File: src/renderer/vulkan/Buffer.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "Buffer.h"

#include "Debug.h"
#include "api/AssertHelper.h"

#include "common/Log.h"
#include "common/ReaperRoot.h"

#include <algorithm>

namespace Reaper
{
namespace
{
    VkBufferUsageFlags BufferUsageToVulkan(u32 usageFlags)
    {
        VkBufferUsageFlags flags = 0;

        flags |= (usageFlags & GPUBufferUsage::TransferSrc) ? VK_BUFFER_USAGE_TRANSFER_SRC_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::TransferDst) ? VK_BUFFER_USAGE_TRANSFER_DST_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::UniformTexelBuffer) ? VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::StorageTexelBuffer) ? VK_BUFFER_USAGE_STORAGE_TEXEL_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::UniformBuffer) ? VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::StorageBuffer) ? VK_BUFFER_USAGE_STORAGE_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::IndexBuffer) ? VK_BUFFER_USAGE_INDEX_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::VertexBuffer) ? VK_BUFFER_USAGE_VERTEX_BUFFER_BIT : 0;
        flags |= (usageFlags & GPUBufferUsage::IndirectBuffer) ? VK_BUFFER_USAGE_INDIRECT_BUFFER_BIT : 0;

        return flags;
    }
} // namespace

GPUBuffer create_buffer(VkDevice device, const char* debug_string, const GPUBufferProperties& input_properties,
                        VmaAllocator& allocator, MemUsage mem_usage)
{
    GPUBufferProperties properties = input_properties;

    // Uniform buffers require extra care on the CPU side.
    // There's a minimum buffer offset we need to take into account.
    // That means there's potentially extra padding between elements regardless
    // of the initial element size.
    if (properties.usage_flags & GPUBufferUsage::UniformBuffer)
    {
        // const u64 minUniformBufferOffsetAlignment =
        // backend.physicalDeviceProperties.limits.minUniformBufferOffsetAlignment;
        const u32 minUniformBufferOffsetAlignment = 0x40; // FIXME
        const u32 stride = std::max(properties.element_size_bytes, minUniformBufferOffsetAlignment);

        properties.stride = stride;
    }
    else
    {
        properties.stride = properties.element_size_bytes;
    }

    const VkBufferCreateInfo bufferInfo = {.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
                                           .pNext = nullptr,
                                           .flags = VK_FLAGS_NONE,
                                           .size = properties.element_count * properties.stride,
                                           .usage = BufferUsageToVulkan(properties.usage_flags),
                                           .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
                                           .queueFamilyIndexCount = 0,
                                           .pQueueFamilyIndices = nullptr};

    VmaAllocationCreateInfo allocInfo = {};

    switch (mem_usage)
    {
    case MemUsage::GPU_Only:
        allocInfo.usage = VMA_MEMORY_USAGE_GPU_ONLY;
        break;
    case MemUsage::CPU_To_GPU:
        allocInfo.usage = VMA_MEMORY_USAGE_CPU_TO_GPU;
        break;
    case MemUsage::GPU_To_CPU:
        allocInfo.usage = VMA_MEMORY_USAGE_GPU_TO_CPU;
        break;
    case MemUsage::CPU_Only:
        allocInfo.usage = VMA_MEMORY_USAGE_CPU_ONLY;
        break;
    default:
        AssertUnreachable();
        break;
    }

    VkBuffer      buffer;
    VmaAllocation allocation;
    Assert(vmaCreateBuffer(allocator, &bufferInfo, &allocInfo, &buffer, &allocation, nullptr) == VK_SUCCESS);

    VulkanSetDebugName(device, buffer, debug_string);

    return GPUBuffer{
        .handle = buffer,
        .allocation = allocation,
        .properties_deprecated = properties,
    };
}

void upload_buffer_data(VkDevice device, const VmaAllocator& allocator, const GPUBuffer& buffer,
                        const GPUBufferProperties& buffer_properties, const void* data, std::size_t size,
                        u32 offset_elements)
{
    u8* writePtr = nullptr;

    Assert(size > 0, "Don't call this function with zero size");

    VmaAllocationInfo allocation_info;
    vmaGetAllocationInfo(allocator, buffer.allocation, &allocation_info);

    const u64 offset_bytes = buffer_properties.stride * offset_elements;

    AssertVk(vkMapMemory(device, allocation_info.deviceMemory, allocation_info.offset + offset_bytes, size,
                         VK_FLAGS_NONE, reinterpret_cast<void**>(&writePtr)));

    Assert(size <= allocation_info.size,
           fmt::format("copy src of size {} on dst of size {}", size, allocation_info.size));

    if (buffer_properties.element_size_bytes == buffer_properties.stride)
    {
        memcpy(writePtr, data, size);
    }
    else
    {
        for (u32 i = 0; i < buffer_properties.element_count; i++)
        {
            const char* data_char = static_cast<const char*>(data);
            memcpy(writePtr + i * buffer_properties.stride, data_char + i * buffer_properties.element_size_bytes,
                   buffer_properties.element_size_bytes);
        }
    }

    vkUnmapMemory(device, allocation_info.deviceMemory);
}

void upload_buffer_data_deprecated(VkDevice device, const VmaAllocator& allocator, const GPUBuffer& buffer,
                                   const void* data, std::size_t size, u32 offset_elements)
{
    return upload_buffer_data(device, allocator, buffer, buffer.properties_deprecated, data, size, offset_elements);
}
} // namespace Reaper

File: src/renderer/vulkan/buffer.zig
// Port of src/renderer/vulkan/Buffer.{h,cpp}

const std = @import("std");
const vk = @import("vulkan");

const vma = @import("vma.zig").c;
const gpu_buffer = @import("../buffer/gpu_buffer.zig");

const GPUBufferProperties = gpu_buffer.GPUBufferProperties;

pub const GPUBuffer = struct {
    handle: vk.Buffer,
    allocation: vma.VmaAllocation,
    properties_deprecated: GPUBufferProperties, // FIXME
};

pub const MemUsage = enum {
    gpu_only,
    cpu_to_gpu,
    gpu_to_cpu,
    cpu_only, // FIXME
};

pub fn bufferUsageToVulkan(usage_flags: gpu_buffer.GPUBufferUsage) vk.BufferUsageFlags {
    return .{
        .transfer_src_bit = usage_flags.transfer_src,
        .transfer_dst_bit = usage_flags.transfer_dst,
        .uniform_texel_buffer_bit = usage_flags.uniform_texel_buffer,
        .storage_texel_buffer_bit = usage_flags.storage_texel_buffer,
        .uniform_buffer_bit = usage_flags.uniform_buffer,
        .storage_buffer_bit = usage_flags.storage_buffer,
        .index_buffer_bit = usage_flags.index_buffer,
        .vertex_buffer_bit = usage_flags.vertex_buffer,
        .indirect_buffer_bit = usage_flags.indirect_buffer,
    };
}

pub fn createBuffer(
    allocator: vma.VmaAllocator,
    input_properties: GPUBufferProperties,
    mem_usage: MemUsage,
) !GPUBuffer {
    var properties = input_properties;

    // Uniform buffers need extra care on the CPU side: there is a minimum
    // buffer offset alignment, which means there can be padding between
    // elements no matter what the element size is.
    if (properties.usage_flags.uniform_buffer) {
        const min_uniform_buffer_offset_alignment: u32 = 0x40; // FIXME
        properties.stride = @max(properties.element_size_bytes, min_uniform_buffer_offset_alignment);
    } else {
        properties.stride = properties.element_size_bytes;
    }

    const buffer_info = vk.BufferCreateInfo{
        .s_type = .buffer_create_info,
        .p_next = null,
        .flags = .{},
        .size = properties.element_count * properties.stride,
        .usage = bufferUsageToVulkan(properties.usage_flags),
        .sharing_mode = .exclusive,
        .queue_family_index_count = 0,
        .p_queue_family_indices = null,
    };

    var alloc_info = std.mem.zeroes(vma.VmaAllocationCreateInfo);
    alloc_info.usage = switch (mem_usage) {
        .gpu_only => vma.VMA_MEMORY_USAGE_GPU_ONLY,
        .cpu_to_gpu => vma.VMA_MEMORY_USAGE_CPU_TO_GPU,
        .gpu_to_cpu => vma.VMA_MEMORY_USAGE_GPU_TO_CPU,
        .cpu_only => vma.VMA_MEMORY_USAGE_CPU_ONLY,
    };

    var buffer: vma.VkBuffer = null;
    var allocation: vma.VmaAllocation = null;

    const result = vma.vmaCreateBuffer(
        allocator,
        @ptrCast(&buffer_info),
        &alloc_info,
        &buffer,
        &allocation,
        null,
    );
    if (result != 0) return error.BufferCreationFailed;

    return .{
        .handle = @enumFromInt(@intFromPtr(buffer)),
        .allocation = allocation,
        .properties_deprecated = properties,
    };
}

pub fn destroyBuffer(allocator: vma.VmaAllocator, buffer: GPUBuffer) void {
    vma.vmaDestroyBuffer(allocator, @ptrFromInt(@intFromEnum(buffer.handle)), buffer.allocation);
}

/// Copies `data` into the buffer, honouring the stride when it differs from
/// the element size.
pub fn uploadBufferData(
    allocator: vma.VmaAllocator,
    buffer: GPUBuffer,
    properties: GPUBufferProperties,
    data: []const u8,
    offset_elements: u32,
) !void {
    var mapped: ?*anyopaque = null;
    if (vma.vmaMapMemory(allocator, buffer.allocation, &mapped) != 0) {
        return error.BufferMapFailed;
    }
    defer vma.vmaUnmapMemory(allocator, buffer.allocation);

    const write_ptr: [*]u8 = @ptrCast(mapped.?);
    const stride = properties.stride;

    std.debug.assert(data.len % properties.element_size_bytes == 0);
    const element_count = data.len / properties.element_size_bytes;

    if (stride == properties.element_size_bytes) {
        const offset_bytes = @as(usize, offset_elements) * stride;
        @memcpy(write_ptr[offset_bytes..][0..data.len], data);
    } else {
        // Strided copy, element by element.
        for (0..element_count) |i| {
            const src = data[i * properties.element_size_bytes ..][0..properties.element_size_bytes];
            const dst_offset = (offset_elements + i) * stride;
            @memcpy(write_ptr[dst_offset..][0..properties.element_size_bytes], src);
        }
    }
}
